include/lsst/afw/cameraGeom/TransformMap.h

```diff
@@ -0,0 +1,213 @@
+/*
+ * LSST Data Management System
+ * Copyright 2014 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#if !defined(LSST_AFW_CAMERAGEOM_TRANSFORMMAP_H)
+#define LSST_AFW_CAMERAGEOM_TRANSFORMMAP_H
+
+#include <vector>
+#include <map>
+#include <unordered_map>
+
+#include "boost/iterator/transform_iterator.hpp"
+#include "astshim/FrameSet.h"
+
+#include "lsst/afw/cameraGeom/CameraSys.h"
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/Transform.h"
+
+namespace lsst {
+namespace afw {
+namespace cameraGeom {
+
+/**
+ * Functor for boost::transform_iterator: given an entry in a std::map or unordered_map, return the key
+ */
+struct GetKey {
+    CameraSys const &operator()(std::pair<const CameraSys, int> const &p) const { return p.first; };
+};
+
+/**
+ * A registry of 2-dimensional coordinate transforms for a specific camera.
+ *
+ * Represents the interrelationships between camera coordinate systems for a
+ * particular camera. It can be seen as a mapping between a pair of CameraSys
+ * and a Transform between them, but does not conform to either the C++ map or
+ * Python dictionary APIs.
+ *
+ * TransformMap supports transforming between any two supported CameraSys using
+ * the @ref transform methods, even if no explicit transform was provided.
+ *
+ * @exceptsafe Unless otherwise specified, all methods guarantee only basic
+ *             exception safety.
+ */
+class TransformMap final {
+public:
+    using Transform = afw::geom::Transform<geom::Point2Endpoint, geom::Point2Endpoint>;
```

Comment:
Defining a `typedef` with the same name as a template can cause a lot of problems (I don't recall exactly what the problems are, but I've always regretted doing it).

I'd also strongly recommend that you have a global typedef for `Transform<Point2Endpoint,Point2Endpoint>`, and just use that throughout (much as we do with e.g. `Point2D`).

Reply:
Could we shorten the Python class names, then?  The simple 2-d to 2-d case is so much more important than all the others that I wouldn't object to calling it `Transform` and using a different name for the template, though e.g. `Transform22` would be acceptable.